nomad/vault.go

```diff
@@ -539,13 +512,55 @@ func (v *vaultClient) renewalLoop() {
 	}
 }
 
+// nextBackoff returns the delay for the next auto renew interval, in seconds.
+// Returns negative value if past expiration
+//
+// It should increaes the amount of backoff each time, with the following rules:
+//
+// * If we have an existing authentication that is going to expire,
+// never back off more than half of the amount of time remaining
+// until expiration (with 1s floor)
+// * Never back off more than 30 seconds multiplied by a random
+// value between 1 and 2
+// * Use randomness so that many clients won't keep hitting Vault
+// at the same time
+func nextBackoff(backoff float64, expiry time.Time) float64 {
+	maxBackoff := time.Until(expiry) / 2
+	if maxBackoff < 0 {
+		return -1
+	}
+
+	switch {
+	case backoff < 5:
+		backoff = 5
+	case backoff >= 24:
+		backoff = 30
+	default:
+		backoff = backoff * 1.25
+	}
+
+	// Add randomness
+	backoff = backoff * (1.0 + rand.Float64())
+
+	if backoff > maxBackoff.Seconds() {
+		backoff = maxBackoff.Seconds()
+	}
+
+	// avoid hammering Vault
+	if backoff < 1 {
+		backoff = 1
```

Comment:
curious if 5 seconds would be a better floor

Reply:
I'm less certain what the right floor is - here we are erring on the side of attempting to renew before expiry; and the difference be making 4 more renewal attempts.

That being said, it was brought up that we want to keep renewing even after token expires until we get an explicit invalid token error from Vault.  In this scenario, it would make sense to set a floor of 5s.